### DISCUSSION TITLE
TSCWG0003 - DevOps and Release Engineering Principles and Practices

### DISCUSSION THREAD

**Discussion by Justus-at-Tazama** [tazama-lf/governance#22] [Community]
This working group is established as a formal working group of the Tazama Technical Steering Committee to review and refine the Tazama DevOps and Release Engineering practices. The working group's ultimate objective is to make contribution to, and deployment of, Tazama easier for contributors and users through automation and appropriate tools.

Working Group Lead: Scott Businge @ Tazama

If you would like to participate, please keep an eye on this discussion for workshop announcements.

**Comment by scott45** [comment]
Thanks Justus.

Some of the things we'd like to prioritise;

- Refining / optimising already existing github action workflows.
- Remove deprecated codeql action v2 in workflows
- Setting up a Dockerhub account as an extension of the release process.
- Docker image builds and dockerhub publish, image scans, docker releases
- Automating packages publishing
- Setup and enable slack notifications to report any success / failure of. builds
- Evaluate other open source tools that will improve our cicd practices 
- Document cicd processes e.g existing workflows and any other undocumented cicd processes we have

**Comment by KennyPaul** [comment]
Link to meeting minutes: https://docs.google.com/document/d/1REeSFeqWBZp0B3gfg4BlrmvxoQPIADYlntfM5jCgNh8/edit 

Link to Google Folder: https://drive.google.com/drive/folders/147_bl4YL7Vkua-ObubcDM_-5VLoKWbMd

**Comment by elnyry-sam-k** [comment]
Following the previous meeting, here's a reference (we don't need to use the exact same way here) or an example for pre-commit hooks to be applied on commits:

https://github.com/mojaloop/central-ledger/blob/033ef631d3add264a11dec8a3f03bcf7932db4f0/package.json#L31

The checks here include lint, dependency checks and basic unit testing to ensure these fundamental checks are already made by contributors before commits can be made.

The dependency check: https://github.com/mojaloop/central-ledger/blob/033ef631d3add264a11dec8a3f03bcf7932db4f0/package.json#L75 ensures dependencies (most of them at least) are up to date and technical debt for future can be minimized.

**Comment by leeivyca** [comment]
Hello I am a new volunteer, looking for ways I can help the project. I have worked in the CI/CD area in the past and I would be interested in learning how you folks are doing it, and maybe I can offer some comments or suggestions. From what I read so far, you seem to be taking a good approach. If your meetings are not too early in the day for a US west coast guy like me, I would be happy to join (I am at leeivyca@gmail.com ).

**Comment by scott45** [comment]
Hello @leeivyca, thanks for showing interest in the project and the cicd rework. We can add you to the next call scheduled for next Thursday, around 9 am USA time.

**Reply by leeivyca** [reply]
Thanks. I looked up your call on https://tazama.org/contact/calendar/ and it appears to be set for 6:00 AM in my time zone (I am on the US west coast, 3 hours behind the east coast). That is pretty early for me. I clicked on the button to request an invite, but I might not be able to wake up that early. If not, I can follow the results by reading the notes.

**Reply by scott45** [reply]
Okay, I'll ask about the possibility of availing you the recording.

**Comment by scott45** [comment]
Current status regarding work progress on some of the set objectives;

1. Move centralised workflows to the new github tazama-lf organisation - In progress, https://github.com/tazama-lf/workflows/tree/main/.github/workflows
2. Document cicd processes e.g existing workflows and any other undocumented cicd processes we have - In progress, PR will be made and added to docs
3. Setting up a Dockerhub account as an extension of the release process. Done - https://hub.docker.com/u/tazamaorg
4. Setup and enable slack notifications to report any success / failure of. builds - Done, Channel name -`#tazama-project-notifications` [Notification message example](https://tazama-lf.slack.com/archives/C07H77XQB3M/p1724054781109079)
5. Refining / optimising already existing github action workflows - In progress e.g. Exclude dependabot prs from running workflows etc
7. Remove deprecated codeql action v2 in workflows (@leeivyca this could be a good first issue for you to play with)

**Comment by scott45** [comment]
## Minutes from the last meeting;

#### Dependant PRs

- What should we consider the frequency for it to be attended to? Justus' qn

**Suggestions from the participants;**

- Document something about dependant that we all can agree on.

- Dependant has 4 levels (critical, medium, high, low). Critical and high immediately. 

- Not including all dependant prs but we need to consider builds.

- Len: devs should bump up the dependancies version.

- Sam: Integrate git hook pre-hook to ensure functionality isn’t breaking. Commit pre-commit hooks - dependency checking  (we have linting / testing). Most of these checks don’t need to run. But definitely good to have. 

**More Questions from the meeting;**

- Is there a gap that we need a documented process for to fix / hot fix

E.g a scenario where we need to figure out for customers urgently. Kenny to provide examples of policies and how it has been addressed (vulnerabilities)

- Do we maybe need to state our SLAs / SLIs - thresholds?? We agreed this might not be practical since we don't offer a PAAS / SAAS. 

Dependency checks shouldn’t be automatic, but contributors can take care of

**PR template additions:** 

- Run the build to ensure nothing breaks. Does it build the work successfully.

- We lack of startup config checking ?? We need to think about this at a certain point

**Action Items from the meeting:**

1. We lack of startup config checking ?? We need to think about this at a certain point
2. What should we consider the frequency for it to be attended to?
3. Integrate git hook pre-hook to ensure functionality isn’t breaking.


**Next meeting agenda suggestion** 

- Ways on how to improve collaboration
- Visual representation of the current cicd process ( what a developer goes through).

**Comment by scott45** [comment]
Current status regarding work progress on some of the set objectives;

- Move centralised workflows to the new github tazama-lf organisation - Done
- Document cicd processes e.g existing workflows and any other undocumented cicd processes we have. Done
- Testing the process of building an image and publishing it to dockerhub. Done, changes to be published across all gtihub orgs by end of next week.
- Remove deprecated codeql action v2 in workflows done by @leeivyca 
- NPM publishing workflow. In progress;
- Adding slack notifications to some workflows (releases, Dockerhub image publishing, npm publishes, Performance benchmarking . In progress

**Comment by scott45** [comment]
**Minutes from the last meeting**

- Software development process presentation from an engineer's pov / daily work (Rodney)

- Group activity (As a contributor, what are your top 5 most frustrating / challenging / annoying manual or automated cicd `tasks`, `processes`, `activities` or `issues`)


**Responses from working group members**;


_**Project / Product Management**_

- I need complete reports, weekly.
- Exception reporting on "CSLAs"
- Reporting for Action (related with OpenSSF Scorecard)

_**DoD**_

- Docs should be updated regularly (Scott)
- Where are the updated guidelines (Scott)
- Need a complete documented step by step process of CI/CD tasks (Scott)

_**PRs**_

- OpenSSF Scorecard (Scott / Kenny)
- More pro-active approach on Dependabot reviews. 
- Transparency and process for when workflows fail

_**Builds**_

- Building the entire platform's docker images (Scott)
- Automated publishing of rule executor + rule processor packages or dockerhub images (Scott)
- Packages built for testing
- Inexhaustive steps for deployment processes (dockerbuild) (Scott)
- Artifact generation

**Comment by scott45** [comment]
### Updates since the last post;


### **Pre 2.1.0 Release (October - December 2024)**


We successfully used some of the completed tasks to automate processes around the 2.1.0 release that happened in December 2024. Some of the achieved targets are listed below


- Remove deprecated codeql action v2 in workflows
- Setting up a Dockerhub account as an extension of the release process.
- Docker image builds and dockerhub image publishing.
- Setup and enable slack notifications to report any success / failure of. builds
- Refining / optimising already existing github action workflows.
- Document cicd processes e.g existing workflows and any other undocumented cicd processes we have


### **2025 Plans**


We have this google doc that we titled [DevOps Setup Current State and Roadmap](https://docs.google.com/document/d/1-kAtvuRExF-CfdXxdhY3NiQf31ZugEznl0OWBcam710/edit?tab=t.0#heading=h.qtk6mtu1pgme)


It highlights the current state the DevOps setup is in from cicd pipelines / development workflows to IAC. The most important section is;


`General Key Areas for Improvement` which includes some of the following points

- Deployment: Update process in existing cloud guides to use published 2.1.0 images on Dockerhub. 
- Secrets Management: Enhance secrets management using an industry standard tool e.g. Infisical, Vault, Doppler etc.
- Automated NPM versioning: The implemented solution during the last release didn’t fully work as expected.
- Security: Tighten and improve on DevSecOps practices.
- Monitoring and Alerting: Integrate Datadog and Prometheus for better observability.
- Dependabot PRs : Use it for updating libraries and figuring out auto merge functionality.
- The end goal is to automate manual processes that are currently being done by developers, SIs, customers etc. Anything that moves in the automation direction is ideal.


### **Roadmap for Improvement**


### `Phase 1: Short-term Goals (1-3 Months)`

- More streamlined Customer support Ops e.g tracking progress in a doc (Prodetect on DO & Nadjib on Azure)
- Ensure CICD working group tasks / activities are clearly defined.
- Defined process for managing Dependabot PRs
- Update guides to use Tazama:2.1.0 images
- Add kubernetes secrets to all deployment.yaml files and secret.yaml files
- Improve documentation and knowledge sharing (DigitalOcean, Azure). Release notes???


### `Phase 2: Mid-term Goals (3-6 Months)`

- Test with ArgoCD as an alternative to Jenkins pipelines.
- Explore open source secrets tools and platforms
- Refine our continuous deployment approach to benefit already running customers.
- Docker images scanning for published artifacts??? 


### `Phase 3: Long-term Goals (6-12 Months)`

- Implement more infrastructure as code for all environments. Ensure we maintain the existing IAC scripts as well might be more useful that adding new ones??? 
- Automate security and compliance checks.

A new [Project Epic](https://github.com/orgs/frmscoe/projects/46) is created in github where tasks/issues are created, assigned and tracked.

Thank you for your input in making Tazama DevOps robust and worldclass.


Please post any suggestions, comments or ideas you have on this thread.
